code/modules/mob/living/carbon/xenomorph/mutation_datum.dm

```diff
@@ -0,0 +1,144 @@
+	.["already_has_veil_mutation"] = has_any_mutation_in_category(xenomorph_user, MUTATION_VEIL)
+
+	.["cost"] = get_mutation_cost(xenomorph_user)
+	.["cost_text"] = (get_mutation_cost(xenomorph_user) > MUTATION_BIOMASS_MAXIMUM || (.["already_has_shell_mutation"] && .["already_has_spur_mutation"] && .["already_has_veil_mutation"])) ? "∞" : .["cost"]
+	.["shell_mutations"] = list()
+	.["spur_mutations"] = list()
+	.["veil_mutations"] = list()
+
+	for(var/datum/mutation_upgrade/mutation AS in xenomorph_user.xeno_caste.mutations)
+		var/list_name = "veil_mutations"
+		if(is_shell_mutation(mutation))
+			list_name = "shell_mutations"
+		else if(is_spur_mutation(mutation))
+			list_name = "spur_mutations"
+		.[list_name] += list(list(
+			"name" = mutation.name,
+			"desc" = mutation.desc,
+			"owned" = has_mutation(xenomorph_user, mutation)
+		))
+
+/datum/mutation_datum/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	if(!isxeno(usr))
+		return
+
+	switch(action)
+		if("purchase")
+			try_purchase_mutation(usr, params["upgrade_name"])
+
+	SStgui.close_user_uis(usr, src)
+
+/// Returns the cost of purchasing a mutation. Cost is based on their caste tier and how many mutations they have so far.
+/datum/mutation_datum/proc/get_mutation_cost(mob/living/carbon/xenomorph/xenomorph_target)
+	var/expected_cost = MUTATION_BIOMASS_THRESHOLD_T4
+	switch(xenomorph_target.xeno_caste.tier)
+		if(XENO_TIER_ONE)
+			expected_cost = MUTATION_BIOMASS_THRESHOLD_T1
+		if(XENO_TIER_TWO)
+			expected_cost = MUTATION_BIOMASS_THRESHOLD_T2
+		if(XENO_TIER_THREE)
+			expected_cost = MUTATION_BIOMASS_THRESHOLD_T3
+	expected_cost += (expected_cost * length(xenomorph_target.owned_mutations))
+	return expected_cost
+
+/// Returns TRUE if the xenomorph has a mutation based on a typepath.
+/datum/mutation_datum/proc/has_mutation(mob/living/carbon/xenomorph/xenomorph_target, datum/mutation_upgrade/mutation_typepath)
+	if(!length(xenomorph_target.owned_mutations))
+		return FALSE
+	for(var/datum/mutation_upgrade/owned_mutation AS in xenomorph_target.owned_mutations)
+		if(!istype(owned_mutation, mutation_typepath))
+			continue
+		return TRUE
+	return FALSE
+
+/// Returns TRUE if the xenomorph has a mutation in a certain category.
+/datum/mutation_datum/proc/has_any_mutation_in_category(mob/living/carbon/xenomorph/xenomorph_target, mutation_category)
+	if(!length(xenomorph_target.owned_mutations))
+		return FALSE
+	for(var/datum/mutation_upgrade/owned_mutation AS in xenomorph_target.owned_mutations)
+		switch(mutation_category)
+			if(MUTATION_SHELL)
+				if(is_shell_mutation(owned_mutation))
+					return TRUE
+			if(MUTATION_SPUR)
+				if(is_spur_mutation(owned_mutation))
+					return TRUE
+			if(MUTATION_VEIL)
+				if(is_veil_mutation(owned_mutation))
+					return TRUE
+	return FALSE
+
+/// Tries to purchase a mutation based on its name. Returns TRUE if the mutation was successfully purchased.
```

Comment:
I dont think a name is a good idea, its more likely to break, I would personally make it text2path the tgui param and do it off typepath